### PR TITLE
[JS/WebGPU] Add GatherBlockQuantized op support

### PR DESCRIPTION
**Pull request by satyajandhyala**
### Description
Add GatherBlockQuantized operator to JSEP.

### Motivation and Context
Gemma model requires this.

### CODE REVIEW

File: js/node/src/tensor_helper.cc [hunk 1]
```diff
@@ -27,7 +27,10 @@ static_assert(ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64 == 13, "definition not consis
 static_assert(ONNX_TENSOR_ELEMENT_DATA_TYPE_COMPLEX64 == 14, "definition not consistent with OnnxRuntime");
 static_assert(ONNX_TENSOR_ELEMENT_DATA_TYPE_COMPLEX128 == 15, "definition not consistent with OnnxRuntime");
 static_assert(ONNX_TENSOR_ELEMENT_DATA_TYPE_BFLOAT16 == 16, "definition not consistent with OnnxRuntime");
-constexpr size_t ONNX_TENSOR_ELEMENT_DATA_TYPE_COUNT = 17;
+static_assert(ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT4X2 == 17, "definition not consistent with OnnxRuntime");
+static_assert(ONNX_TENSOR_ELEMENT_DATA_TYPE_INT4X2 == 18, "definition not consistent with OnnxRuntime");
```
Comment (satyajandhyala) [inline js/node/src/tensor_helper.cc:31]: Reverting these changes

File: js/web/lib/wasm/jsep/webgpu/ops/gather-block-quantized.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+          }
+        })()};
+        let dequantized_data = ${tensorTypeToWsglValueType(outputType)}(quantized_data - zero_point) * scale;
+        ${output.setByOffset('global_idx', 'dequantized_data')};
+    }`;
+  };
+  return {
+    name: 'GatherBlockQuantized',
+    shaderCache: {
+      hint: `${attributes.cacheKey};${inputs.map((input, _) => input.dims.join('_')).join(';')}`,
```
Comment (guschmue) [inline js/web/lib/wasm/jsep/webgpu/ops/gather-block-quantized.ts:168]: inputs[1] is input_ids - we should avoid those dims from being known in the shader.

**Review by guschmue — CHANGES_REQUESTED** [review]

File: js/web/lib/wasm/jsep/webgpu/ops/gather-block-quantized.ts (hunk 2 above)
Comment (satyajandhyala) [reply js/web/lib/wasm/jsep/webgpu/ops/gather-block-quantized.ts:168]: We are only adding the shape of the indices input to the hint. Not the indices themselves. If the shape of the indices changes then the output shape also would change. Please confirm.

Comment (satyajandhyala) [reply js/web/lib/wasm/jsep/webgpu/ops/gather-block-quantized.ts:168]: A part of the output_indices goes to indexing input indices and the rest goes to indexing input combined with index from input_indices.

Comment (satyajandhyala) [reply js/web/lib/wasm/jsep/webgpu/ops/gather-block-quantized.ts]: Removed input[1] from cacheKey

**Review by guschmue — APPROVED** [review]
works for me. There is some wired issue when using this with gemma-2-2b but I don't think this is related to this PR.

**Pull request merged**